### ISSUE TITLE
Questions

### ISSUE THREAD

**Issue by emilbaekdahl** [MIRALab-USTC/KGE-HAKE#4]
Hi,

First of all, thanks for sharing the code behind your paper! I have a few questions; mainly about how the details in the paper maps to the code.

1. My main concern is how a semantic hierarchies is translated into the embedding. How do we go from text values from the knowledge graphs to an embedding with modulus values that obey the hierarchy; for instance "dog" and "mammal" gets embedded with modulus values 2 and 1? Is it a consequence of the score function? 

2. In the code, we have the parameters `modulus_weight` and `phase_weight` that are used to [multiply with the modulus and phase score](https://github.com/MIRALab-USTC/KGE-HAKE/blob/63b9fec5b2f871cc5d99c25de32f2cb905d51978/codes/models.py#L362). In the paper, however, we only multiply with a weight in the phase part (the lambda parameter) as seen below.
![Screenshot 2020-05-11 at 11 49 30](https://user-images.githubusercontent.com/21752915/81548372-7b79db00-937d-11ea-8038-81e7bf2a90ed.png)
So the question is, are we missing a weight in the modulus part? If not, where does the above lambda parameter map to in the code?

3. I can not find a description of the `embedding_range` parameter in the paper. Is there an explanation behind [its definition](https://github.com/MIRALab-USTC/KGE-HAKE/blob/63b9fec5b2f871cc5d99c25de32f2cb905d51978/codes/models.py#L309)? Furthermore, is the purpose of the [epsilon value](https://github.com/MIRALab-USTC/KGE-HAKE/blob/63b9fec5b2f871cc5d99c25de32f2cb905d51978/codes/models.py#L301) just to get a minimum value for the embedding range? And why `2.0`?

4. The initialisation of the `HAKE` and `ModE` classes seems to differ in how the relation embedding is initialised. The `HAKE` initialises the modulus and bias to [ones](https://github.com/MIRALab-USTC/KGE-HAKE/blob/63b9fec5b2f871cc5d99c25de32f2cb905d51978/codes/models.py#L327) and [zeroes](https://github.com/MIRALab-USTC/KGE-HAKE/blob/63b9fec5b2f871cc5d99c25de32f2cb905d51978/codes/models.py#L331) respectively. What is the purpose of this?

I hope the questions make sense. Thanks in advance!

**Comment by zhanqiuzhang** [comment]
Thanks for your interest in our work.

1. We believe that it is a consequence of the score function and the modulus part $\|h\circ r-t\|$ can model different hierarchies. In our paper, we conduct extensive experiments to verify the conclusion. Moreover, Figure 2 shows that even only using modulus information, ModE can embed relations---that connect entities at lower levels to entities at higher levels---to modulus values less than one. However, it still an empirical result and requires further theoretical analysis. We think it is mainly due to the one-to-many/many-to-one patterns in semantic hierarchies (e.g. both "dog" and "cat" are "mammal") and the optimization process with gradient descent.
2. We have mentioned the parameter in the Training Protocol part of our paper. These two formulations are equivalent when we score a triple, while we find that adding the `modulus_weight ` makes searching the best hyperparameters easier.
3. We initialize some of the embedding parameters with values from the uniform distribution, so we need to set the lower bound and upper bound of the distribution. The `embedding_range` in the code is a hyperparameter to determine the bounds. Yes, the purpose of the `epsilon` value is just to get a value for the embedding range.
4. (a). For $[r_m]_i$ in HAKE, we have either $0 < [r_m]_i < 1$, $[r_m]_i = 1$ or $[r_m]_i > 1$. Therefore, we intuitively initialize $[r_m]_i$ with 1.0 such that all $[r_m]_i$ could converge to their optimals with less costs.
(b). For ModE, $[r]_i$ can take negative values, so we uniformly initialize $[r]_i$ within a small range centered at zero.

Hope it helps.
